src/preamble.js

```diff
@@ -1001,7 +1001,7 @@ function createWasm() {
     assert(wasmTable, 'table not found in wasm exports');
 #endif
 
-#if AUDIO_WORKLET
+#if AUDIO_WORKLET && !(EXPORT_ES6 && ASSERTIONS)  // If EXPORT_ES6, we may have already done this.
```

Comment:
Can you explain this some more? In particular,
* Where would we have possibly done it before?
* Why is it a problem if we've done it before?
* Why is this an issue with ASSERTIONS specifically? (That seems surprising as that mode should only add additional checks, not affect how many times actual non-assertion logic happens.)

Reply:
OK, here goes.

## Does `ASSERTIONS` specifically matter?
Surprisingly, yes.

Currently, with `ASSERTIONS`, the following is generated in the output JS, which turns `Module.wasmTable` into a read-only getter:

```js
function unexportedRuntimeSymbol(sym) { [ ... ] }

[ ... ]

var unexportedSymbols = [ ..., 'wasmTable', ... ];
unexportedSymbols.forEach(unexportedRuntimeSymbol);
```

Afterwards, trying to read `Module.wasmTable` during module init (here "module" refers to the JS ES6 module, and "module init" to the async module + binary .wasm loading and compilation step) will cause module init to fail with a message that `wasmTable` hasn't been explicitly exported. This makes sense for normal ES6 modules to prevent the user from doing probably unintended things.

The issue stems from Audio Worklet ES6 modules relying upon assigning `Module['wasmTable'] = wasmTable` to make `wasmTable` available to the bootstrapping code in the separate .aw.js file loaded into the Audio Worklet (i.e. `audioWorklet.addModule('test.aw.js')`. While we don't try to read `Module.wasmTable`, we do try to assign to it, and the preexistence of the read-only getter means that the assignment&mdash;and module init&mdash;will fail.

The failure does not occur with `ASSERTIONS=0`, nor after adding `wasmTable` to `EXPORTED_RUNTIME_METHODS` (as suggested in the error message emitted by the read-only getter).

https://github.com/emscripten-core/emscripten/blob/ec2ce4ce1d1596e3c45bdcd0a8e7cdeb0fe4ab3f/src/runtime_debug.js#L87

## Did we `Module['wasmTable'] = wasmTable` already?
Adding `wasmTable` to `EXPORTED_RUNTIME_METHODS` is, by itself, sufficient for doing the `Module['wasmTable'] = wasmTable` assignment. Here's src/preamble.js:1004 with a bit more leading context:

```js 
#if '$wasmTable' in addedLibraryItems && !RELOCATABLE
    wasmTable = wasmExports['__indirect_function_table'];
    {{{ receivedSymbol('wasmTable') }}}
#if ASSERTIONS && !PURE_WASI
    assert(wasmTable, 'table not found in wasm exports');
#endif

#if AUDIO_WORKLET
    // If we are in the audio worklet environment, we can only access the Module object
    // and not the global scope of the main JS script. Therefore we need to export
    // all functions that the audio worklet scope needs onto the Module object.
    Module['wasmTable'] = wasmTable;
#endif
```

`{{{ receivedSymbol('wasmTable') }}}` becomes `Module['wasmTable'] = wasmTable;` in the generated JS.

## Does it matter if we `Module['wasmTable'] = wasmTable` again?
Not in this case, except for a little code duplication.

Related, but I've since noticed that adding `wasmTable` to `EXPORTED_RUNTIME_METHODS` creates that exact assignment in the postamble:
https://github.com/emscripten-core/emscripten/blob/a9ebfb74aef9090ca42b3ebf493722bd1ed24854/src/postamble.js#L27

This seems like the way exported runtime methods are "supposed" to be exported, and the current assignments were written without that in mind. That being the case, perhaps something like the following would be better:

```js
// src/parseTools.js

function hasExportedRuntimeMethod(sym) {
  return EXPORTED_RUNTIME_METHODS.includes(sym);
}
```

```js
// src/preamble.js

#if '$wasmTable' in addedLibraryItems && !RELOCATABLE
    wasmTable = wasmExports['__indirect_function_table'];
#if !hasExportedRuntimeMethod('wasmTable')
    {{{ receivedSymbol('wasmTable') }}}
#endif
#if ASSERTIONS && !PURE_WASI
    assert(wasmTable, 'table not found in wasm exports');
#endif

#if AUDIO_WORKLET && !hasExportedRuntimeMethod('wasmTable')
    // If we are in the audio worklet environment, we can only access the Module object
    // and not the global scope of the main JS script. Therefore we need to export
    // all functions that the audio worklet scope needs onto the Module object.

    Module['wasmTable'] = wasmTable;
#endif
```
Thoughts?

Reply:
I think the issue is that wasmTable wasn't always something that could be exported via EXPORTED_RUNTIME_METHODS.

Now that it is, can you remove the `Module['wasmTable'] = wasmTable` code block above and move that comment into `link.py` where it gets added to `EXPORTED_RUNTIME_METHODS`?

You would also need to make that addion no longer conditional on `EXPORT_ES6` or `ASSERTIONS`, which I think is cleaner.

Reply:
> Now that it is, can you remove the `Module['wasmTable'] = wasmTable` code block above and move that comment into `link.py` where it gets added to `EXPORTED_RUNTIME_METHODS`?
Certainly. Should I also remove the template call to `{{{ receivedSymbol('wasmTable') }}}`?

For that matter, is `receivedSymbol()` even necessary for exporting `wasmExports` and `wasmMemory` anymore?

> You would also need to make that addion no longer conditional on `EXPORT_ES6` or `ASSERTIONS`, which I think is cleaner.

Agreed, will do. I initially had it that way but made it conditional because ES6 modules are my use case and I didn't want to risk affecting non-ES6 Audio Worklets.

Reply:
Actually I tried to do this change myself a few weeks back an ended up going down a rabit hole.  I think this change is right way to go but it might be a complicated change.

Reply:
So `{{{ receivedSymbol('wasmTable') }}}` should stay? Of `wasmExports`, `wasmMemory`, and `wasmTable`, `wasmTable` looks the least rabbit-holey of the bunch.

Reply:
I think all the usages of `receivedSymbol` should probably stay yes.

The change I was trying to make was to completely remove `receivedSymbol` and use the same mechanism for those symbols at we use for exported wasm functions and globals.  That proved tricky though so I gave up for now.

Reply:
Got it.

Reply:
Revised, rebased, and pushed.

I noticed that the same read-only getter issue was occurring with `MINIMAL_RUNTIME`, and it was affecting non-`EXPORT_ES6` builds as well. It seems several automated tests for Audio Worklets are passing when they shouldn't :raised_eyebrow:

I'm not sure where to start with looking into that, but the change is covered by an existing test. `browser.test_audio_worklet_minimal_runtime` was wrongly passing before, and it is correctly passing now. So I added the fix to the PR.

FWIW, I did manually test `MINIMAL_RUNTIME` + `EXPORT_ES6` in combination. It worked.

Reply:
Thank you for this work, @kangtastic . The `URL is not defined` bug has been driving me nuts for a few days.